File: sdk/rust/src/instructions.rs
//! Vault program `Instruction` builders (account order matches `program/src/instructions/*` and the TS SDK).

use solana_instruction::{AccountMeta, Instruction};
use solana_pubkey::Pubkey;

use crate::constants::{
   ASSOCIATED_TOKEN_PROGRAM_ID, DEFAULT_VAULT_PROGRAM_ID, SPL_TOKEN_PROGRAM_ID, SYSTEM_PROGRAM_ID,
   SYSVAR_CLOCK_ID, SYSVAR_INSTRUCTIONS_ID,
};
use crate::error::VaultSdkError;

/// On-chain instruction indices (`program/src/instructions/mod.rs`).
#[repr(u8)]
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum VaultInstructionKind {
   CreateUserVault = 0,
   DepositUserVault = 1,
   UpdateUserVaultDelegate = 2,
   WithdrawUserVault = 3,
   WithdrawUserVaultNative = 4,
   AppIx = 5,
   CpiEntry = 6,
   CpiEntryNative = 7,
   CloseVaultAta = 8,
   CloseUserVault = 9,
}

fn require_positive_amount(amount: u64) -> Result<(), VaultSdkError> {
   if amount == 0 {
      Err(VaultSdkError::InvalidAmount)
   } else {
      Ok(())
   }
}

fn require_cpi_entry_amounts(amount_native: u64, amount: u64) -> Result<(), VaultSdkError> {
   if amount_native == 0 && amount == 0 {
      Err(VaultSdkError::InvalidAmount)
   } else {
      Ok(())
   }
}

/// `CreateUserVault` — 5 accounts: owner, user_vault_pda, app, delegate, system. Data: discriminator + `delegate_expires` u32 LE (Unix seconds; `u32::MAX` = no practical expiry).
pub fn create_user_vault_ix(
   program_id: &Pubkey,
   owner: &Pubkey,
   user_vault_pda: &Pubkey,
   app_address: &Pubkey,
   delegate: &Pubkey,
   system_program: Option<&Pubkey>,
   delegate_expires: u32,
) -> Instruction {
   let system = system_program.copied().unwrap_or(SYSTEM_PROGRAM_ID);
   let mut data = Vec::with_capacity(5);
   data.push(VaultInstructionKind::CreateUserVault as u8);
   data.extend_from_slice(&delegate_expires.to_le_bytes());
   Instruction {
      program_id: *program_id,
      accounts: vec![
         AccountMeta::new(*owner, true),
         AccountMeta::new(*user_vault_pda, false),
         AccountMeta::new_readonly(*app_address, false),
         AccountMeta::new_readonly(*delegate, false),
         AccountMeta::new_readonly(system, false),
      ],
      data,
   }
}

/// `DepositUserVault` — 9 accounts; `amount` is little-endian `u64` after discriminator.
pub fn deposit_user_vault_ix(
   program_id: &Pubkey,
   owner: &Pubkey,
   user_vault_pda: &Pubkey,
   user_vault_ata: &Pubkey,
   app_address: &Pubkey,
   source_ata: &Pubkey,
   mint: &Pubkey,
   token_program: &Pubkey,
   amount: u64,
   system_program: Option<&Pubkey>,
   associated_token_program: Option<&Pubkey>,
) -> Result<Instruction, VaultSdkError> {
   require_positive_amount(amount)?;
   let system = system_program.copied().unwrap_or(SYSTEM_PROGRAM_ID);
   let ata_program = associated_token_program.copied().unwrap_or(ASSOCIATED_TOKEN_PROGRAM_ID);
   let mut data = Vec::with_capacity(9);
   data.push(VaultInstructionKind::DepositUserVault as u8);
   data.extend_from_slice(&amount.to_le_bytes());
   Ok(Instruction {
      program_id: *program_id,
      accounts: vec![
         AccountMeta::new(*owner, true),
         AccountMeta::new(*user_vault_pda, false),
         AccountMeta::new(*user_vault_ata, false),
         AccountMeta::new_readonly(*app_address, false),
         AccountMeta::new(*source_ata, false),
         AccountMeta::new_readonly(*mint, false),
         AccountMeta::new_readonly(system, false),
         AccountMeta::new_readonly(*token_program, false),
         AccountMeta::new_readonly(ata_program, false),
      ],
      data,
   })
}

/// `UpdateUserVaultDelegate` — 4 accounts. Data: discriminator + `delegate_expires` u32 LE (Unix seconds; `u32::MAX` = no practical expiry).
pub fn update_user_vault_delegate_ix(
   program_id: &Pubkey,
   owner: &Pubkey,
   user_vault_pda: &Pubkey,
   app_address: &Pubkey,
   delegate: &Pubkey,
   delegate_expires: u32,
) -> Instruction {
   let mut data = Vec::with_capacity(5);
   data.push(VaultInstructionKind::UpdateUserVaultDelegate as u8);
   data.extend_from_slice(&delegate_expires.to_le_bytes());
   Instruction {
      program_id: *program_id,
      accounts: vec![
         AccountMeta::new(*owner, true),
         AccountMeta::new(*user_vault_pda, false),
         AccountMeta::new_readonly(*app_address, false),
         AccountMeta::new_readonly(*delegate, false),
      ],
      data,
   }
}

/// `WithdrawUserVault` — 7 accounts.
pub fn withdraw_user_vault_ix(
   program_id: &Pubkey,
   owner: &Pubkey,
   user_vault_pda: &Pubkey,
   user_vault_ata: &Pubkey,
   app_address: &Pubkey,
   dest_ata: &Pubkey,
   mint: &Pubkey,
   token_program: &Pubkey,
   amount: u64,
) -> Result<Instruction, VaultSdkError> {
   require_positive_amount(amount)?;
   let mut data = Vec::with_capacity(9);
   data.push(VaultInstructionKind::WithdrawUserVault as u8);
   data.extend_from_slice(&amount.to_le_bytes());
   Ok(Instruction {
      program_id: *program_id,
      accounts: vec![
         AccountMeta::new(*owner, true),
         AccountMeta::new_readonly(*user_vault_pda, false),
         AccountMeta::new(*user_vault_ata, false),
         AccountMeta::new_readonly(*app_address, false),
         AccountMeta::new(*dest_ata, false),
         AccountMeta::new_readonly(*mint, false),
         AccountMeta::new_readonly(*token_program, false),
      ],
      data,
   })
}

/// `WithdrawUserVaultNative` — 3 accounts: owner (signer, writable), user_vault_pda (writable), app. Data: discriminator + `amount` u64 LE.
pub fn withdraw_user_vault_native_ix(
   program_id: &Pubkey,
   owner: &Pubkey,
   user_vault_pda: &Pubkey,
   app_address: &Pubkey,
   amount: u64,
) -> Result<Instruction, VaultSdkError> {
   require_positive_amount(amount)?;
   let mut data = Vec::with_capacity(9);
   data.push(VaultInstructionKind::WithdrawUserVaultNative as u8);
   data.extend_from_slice(&amount.to_le_bytes());
   Ok(Instruction {
      program_id: *program_id,
      accounts: vec![
         AccountMeta::new(*owner, true),
         AccountMeta::new(*user_vault_pda, false),
         AccountMeta::new_readonly(*app_address, false),
      ],
      data,
   })
}

/// Fixed accounts for `AppIx` (delegate signer + vault context). Append inner CPI metas after these.
pub struct AppIxFixedAccounts<'a> {
   pub delegate: &'a Pubkey,
   pub owner: &'a Pubkey,
   pub user_vault_pda: &'a Pubkey,
   pub app_address: &'a Pubkey,
}

/// `AppIx` — 4 fixed accounts, then accounts for the inner app instruction. `inner_instruction_data` is passed through to the CPI as-is (no length prefix).
pub fn app_ix_ix(
   program_id: &Pubkey,
   fixed: AppIxFixedAccounts,
   inner_accounts: Vec<AccountMeta>,
   inner_instruction_data: &[u8],
) -> Instruction {
   let mut data = Vec::with_capacity(1 + inner_instruction_data.len());
   data.push(VaultInstructionKind::AppIx as u8);
   data.extend_from_slice(inner_instruction_data);
   let mut accounts = vec![
      AccountMeta::new(*fixed.delegate, true),
      AccountMeta::new_readonly(*fixed.owner, false),
      AccountMeta::new_readonly(*fixed.user_vault_pda, false),
      AccountMeta::new_readonly(*fixed.app_address, false),
   ];
   accounts.extend(inner_accounts);
   Instruction {
      program_id: *program_id,
      accounts,
      data,
   }
}

/// Builds the 11 `AccountMeta` entries for `CpiEntry` (`program/src/instructions/cpi_entry.rs`).
/// Writable flags for vault PDA, vault ATA, lamports destination, and destination ATA follow the amounts you pass in
/// (`amount_native` / `amount` must match the serialized instruction data).
pub fn cpi_entry_account_metas(
   delegate: &Pubkey,
   owner: &Pubkey,
   user_vault_pda: &Pubkey,
   user_vault_ata: &Pubkey,
   app_address: &Pubkey,
   lamports_dest: &Pubkey,
   dest_ata: &Pubkey,
   mint: &Pubkey,
   token_program: &Pubkey,
   amount_native: u64,
   amount: u64,
   instructions_sysvar: Option<&Pubkey>,
   clock_sysvar: Option<&Pubkey>,
) -> Vec<AccountMeta> {
   let w_native = amount_native > 0;
   let w_spl = amount > 0;
   let ixs = instructions_sysvar.copied().unwrap_or(SYSVAR_INSTRUCTIONS_ID);
   let clock = clock_sysvar.copied().unwrap_or(SYSVAR_CLOCK_ID);
   vec![
      AccountMeta::new(*delegate, true),
      AccountMeta::new_readonly(*owner, false),
      if w_native {
         AccountMeta::new(*user_vault_pda, false)
      } else {
         AccountMeta::new_readonly(*user_vault_pda, false)
      },
      if w_spl {
         AccountMeta::new(*user_vault_ata, false)
      } else {
         AccountMeta::new_readonly(*user_vault_ata, false)
      },
      AccountMeta::new_readonly(*app_address, false),
      if w_native {
         AccountMeta::new(*lamports_dest, false)
      } else {
         AccountMeta::new_readonly(*lamports_dest, false)
      },
      if w_spl {
         AccountMeta::new(*dest_ata, false)
      } else {
         AccountMeta::new_readonly(*dest_ata, false)
      },
      AccountMeta::new_readonly(*mint, false),
      AccountMeta::new_readonly(*token_program, false),
      AccountMeta::new_readonly(ixs, false),
      AccountMeta::new_readonly(clock, false),
   ]
}

/// `CpiEntry` (vault CPI path) — data is discriminator + `amount_native` u64 LE + `amount` u64 LE (SPL token amount).
pub fn cpi_entry_ix(
   program_id: &Pubkey,
   delegate: &Pubkey,
   owner: &Pubkey,
   user_vault_pda: &Pubkey,
   user_vault_ata: &Pubkey,
   app_address: &Pubkey,
   lamports_dest: &Pubkey,
   dest_ata: &Pubkey,
   mint: &Pubkey,
   token_program: &Pubkey,
   amount_native: u64,
   amount: u64,
   instructions_sysvar: Option<&Pubkey>,
   clock_sysvar: Option<&Pubkey>,
) -> Result<Instruction, VaultSdkError> {
   require_cpi_entry_amounts(amount_native, amount)?;
   let accounts = cpi_entry_account_metas(
      delegate,
      owner,
      user_vault_pda,
      user_vault_ata,
      app_address,
      lamports_dest,
      dest_ata,
      mint,
      token_program,
      amount_native,
      amount,
      instructions_sysvar,
      clock_sysvar,
   );
   let mut data = Vec::with_capacity(17);
   data.push(VaultInstructionKind::CpiEntry as u8);
   data.extend_from_slice(&amount_native.to_le_bytes());
   data.extend_from_slice(&amount.to_le_bytes());
   Ok(Instruction {
      program_id: *program_id,
      accounts,
      data,
   })
}

/// Seven accounts for `CpiEntryNative` (`program/src/instructions/cpi_entry_native.rs`).
pub fn cpi_entry_native_account_metas(
   delegate: &Pubkey,
   owner: &Pubkey,
   user_vault_pda: &Pubkey,
   app_address: &Pubkey,
   lamports_dest: &Pubkey,
   instructions_sysvar: Option<&Pubkey>,
   clock_sysvar: Option<&Pubkey>,
) -> Vec<AccountMeta> {
   let ixs = instructions_sysvar.copied().unwrap_or(SYSVAR_INSTRUCTIONS_ID);
   let clock = clock_sysvar.copied().unwrap_or(SYSVAR_CLOCK_ID);
   vec![
      AccountMeta::new(*delegate, true),
      AccountMeta::new_readonly(*owner, false),
      AccountMeta::new(*user_vault_pda, false),
      AccountMeta::new_readonly(*app_address, false),
      AccountMeta::new(*lamports_dest, false),
      AccountMeta::new_readonly(ixs, false),
      AccountMeta::new_readonly(clock, false),
   ]
}

/// `CpiEntryNative` (vault CPI path) — data is discriminator + `amount_native` u64 LE.
pub fn cpi_entry_native_ix(
   program_id: &Pubkey,
   delegate: &Pubkey,
   owner: &Pubkey,
   user_vault_pda: &Pubkey,
   app_address: &Pubkey,
   lamports_dest: &Pubkey,
   amount_native: u64,
   instructions_sysvar: Option<&Pubkey>,
   clock_sysvar: Option<&Pubkey>,
) -> Result<Instruction, VaultSdkError> {
   require_positive_amount(amount_native)?;
   let accounts = cpi_entry_native_account_metas(
      delegate,
      owner,
      user_vault_pda,
      app_address,
      lamports_dest,
      instructions_sysvar,
      clock_sysvar,
   );
   let mut data = Vec::with_capacity(9);
   data.push(VaultInstructionKind::CpiEntryNative as u8);
   data.extend_from_slice(&amount_native.to_le_bytes());
   Ok(Instruction {
      program_id: *program_id,
      accounts,
      data,
   })
}

/// `CloseVaultAta` — 7 accounts.
pub fn close_vault_ata_ix(
   program_id: &Pubkey,
   owner: &Pubkey,
   user_vault_pda: &Pubkey,
   app_address: &Pubkey,
   user_vault_ata: &Pubkey,
   destination: &Pubkey,
   mint: &Pubkey,
   token_program: &Pubkey,
) -> Instruction {
   Instruction {
      program_id: *program_id,
      accounts: vec![
         AccountMeta::new(*owner, true),
         AccountMeta::new(*user_vault_pda, false),
         AccountMeta::new_readonly(*app_address, false),
         AccountMeta::new(*user_vault_ata, false),
         AccountMeta::new(*destination, false),
         AccountMeta::new_readonly(*mint, false),
         AccountMeta::new_readonly(*token_program, false),
      ],
      data: vec![VaultInstructionKind::CloseVaultAta as u8],
   }
}

/// `CloseUserVault` — 3 accounts.
pub fn close_user_vault_ix(program_id: &Pubkey, owner: &Pubkey, user_vault_pda: &Pubkey, app_address: &Pubkey) -> Instruction {
   Instruction {
      program_id: *program_id,
      accounts: vec![
         AccountMeta::new(*owner, true),
         AccountMeta::new(*user_vault_pda, false),
         AccountMeta::new_readonly(*app_address, false),
      ],
      data: vec![VaultInstructionKind::CloseUserVault as u8],
   }
}

// --- Convenience wrappers using [`DEFAULT_VAULT_PROGRAM_ID`] ---

/// Same as [`create_user_vault_ix`] with `program_id` = [`DEFAULT_VAULT_PROGRAM_ID`].
pub fn create_user_vault_ix_default(
   owner: &Pubkey,
   user_vault_pda: &Pubkey,
   app_address: &Pubkey,
   delegate: &Pubkey,
   system_program: Option<&Pubkey>,
   delegate_expires: u32,
) -> Instruction {
   create_user_vault_ix(
      &DEFAULT_VAULT_PROGRAM_ID,
      owner,
      user_vault_pda,
      app_address,
      delegate,
      system_program,
      delegate_expires,
   )
}

/// Same as [`deposit_user_vault_ix`] with `program_id` = [`DEFAULT_VAULT_PROGRAM_ID`].
pub fn deposit_user_vault_ix_default(
   owner: &Pubkey,
   user_vault_pda: &Pubkey,
   user_vault_ata: &Pubkey,
   app_address: &Pubkey,
   source_ata: &Pubkey,
   mint: &Pubkey,
   token_program: &Pubkey,
   amount: u64,
   system_program: Option<&Pubkey>,
   associated_token_program: Option<&Pubkey>,
) -> Result<Instruction, VaultSdkError> {
   deposit_user_vault_ix(
      &DEFAULT_VAULT_PROGRAM_ID,
      owner,
      user_vault_pda,
      user_vault_ata,
      app_address,
      source_ata,
      mint,
      token_program,
      amount,
      system_program,
      associated_token_program,
   )
}

/// Uses classic SPL Token as `token_program` for [`deposit_user_vault_ix_default`].
pub fn deposit_user_vault_ix_default_spl_token(
   owner: &Pubkey,
   user_vault_pda: &Pubkey,
   user_vault_ata: &Pubkey,
   app_address: &Pubkey,
   source_ata: &Pubkey,
   mint: &Pubkey,
   amount: u64,
   system_program: Option<&Pubkey>,
   associated_token_program: Option<&Pubkey>,
) -> Result<Instruction, VaultSdkError> {
   deposit_user_vault_ix_default(
      owner,
      user_vault_pda,
      user_vault_ata,
      app_address,
      source_ata,
      mint,
      &SPL_TOKEN_PROGRAM_ID,
      amount,
      system_program,
      associated_token_program,
   )
}

File: sdk/ts/src/instructions.ts
import { getAddressEncoder, getProgramDerivedAddress, type Address } from '@solana/addresses';
import { AccountRole, type AccountMeta, type Instruction } from '@solana/instructions';

import {
   ASSOCIATED_TOKEN_PROGRAM_ADDRESS,
   SYSTEM_PROGRAM_ADDRESS,
   SYSVAR_CLOCK_ADDRESS,
   SYSVAR_INSTRUCTIONS_ADDRESS,
   SYSVAR_RENT_ADDRESS,
   USER_VAULT_SEED,
   VAULT_PROGRAM_ADDRESS,
} from './constants';
import {
   assertCpiEntryAmounts,
   assertU32DelegateExpires,
   assertU64Amount,
   getVaultInstructionDataEncoder,
} from './codex';
import type {
   AppIxInput,
   CpiEntryInput,
   CpiEntryNativeInput,
   CreateUserVaultInput,
   DecodedVaultInstruction,
   DepositUserVaultInput,
   UpdateUserVaultDelegateInput,
   WithdrawUserVaultInput,
   WithdrawUserVaultNativeInput,
} from './types';

/**
 * `cpi_entry` / `cpi_entry_native` are invoked via CPI from the registered app program (instructions + clock sysvars).
 * Builders match `cpi_entry.rs` / `cpi_entry_native.rs`.
 * `app_ix` matches `app_ix.rs`: five fixed accounts (ending with clock sysvar), then inner metas.
 */

const textEncoder = new TextEncoder();

function encodeVaultIx(ix: DecodedVaultInstruction): Uint8Array {
   return new Uint8Array(getVaultInstructionDataEncoder().encode(ix));
}

export async function deriveUserVaultPda(
   programAddress: Address,
   owner: Address,
   appAddress: Address,
): Promise<readonly [Address, number]> {
   const addrEnc = getAddressEncoder();
   const [pda, bump] = await getProgramDerivedAddress({
      programAddress,
      seeds: [textEncoder.encode(USER_VAULT_SEED), addrEnc.encode(owner), addrEnc.encode(appAddress)],
   });
   return [pda, bump] as const;
}

/** SPL associated token address for wallet = `userVaultPda`, explicit classic or Token-2022 program. */
export async function deriveUserVaultAtaAddress(
   userVaultPda: Address,
   mint: Address,
   tokenProgram: Address,
): Promise<Address> {
   const addrEnc = getAddressEncoder();
   const [ata] = await getProgramDerivedAddress({
      programAddress: ASSOCIATED_TOKEN_PROGRAM_ADDRESS,
      seeds: [addrEnc.encode(userVaultPda), addrEnc.encode(tokenProgram), addrEnc.encode(mint)],
   });
   return ata;
}

/** SPL ATA for a wallet owner (same PDA layout as the vault ATA, but authority = `owner`). */
export async function deriveAtaAddress(
   owner: Address,
   mint: Address,
   tokenProgram: Address,
): Promise<Address> {
   const addrEnc = getAddressEncoder();
   const [ata] = await getProgramDerivedAddress({
      programAddress: ASSOCIATED_TOKEN_PROGRAM_ADDRESS,
      seeds: [addrEnc.encode(owner), addrEnc.encode(tokenProgram), addrEnc.encode(mint)],
   });
   return ata;
}

export function getCreateUserVaultInstruction(input: {
   accounts: Readonly<{
      owner: Address;
      userVaultPda: Address;
      appAddress: Address;
      delegate: Address;
   }>;
   data: CreateUserVaultInput;
}): Instruction {
   assertU32DelegateExpires(input.data.delegateExpires, 'createUserVault delegateExpires');
   return {
      programAddress: VAULT_PROGRAM_ADDRESS,
      accounts: [
         { address: input.accounts.owner, role: AccountRole.WRITABLE_SIGNER },
         { address: input.accounts.userVaultPda, role: AccountRole.WRITABLE },
         { address: input.accounts.appAddress, role: AccountRole.READONLY },
         { address: input.accounts.delegate, role: AccountRole.READONLY },
         { address: SYSVAR_RENT_ADDRESS, role: AccountRole.READONLY },
         { address: SYSTEM_PROGRAM_ADDRESS, role: AccountRole.READONLY },
      ] as const satisfies readonly AccountMeta[],
      data: encodeVaultIx({
         kind: 'createUserVault',
         delegateExpires: input.data.delegateExpires,
      }),
   };
}

export function getDepositUserVaultInstruction(input: {
   accounts: Readonly<{
      owner: Address;
      userVaultPda: Address;
      userVaultAta: Address;
      appAddress: Address;
      sourceAta: Address;
      mint: Address;
      tokenProgram: Address;
   }>;
   data: DepositUserVaultInput;
}): Instruction {
   assertU64Amount(input.data.amount, 'deposit amount');
   return {
      programAddress: VAULT_PROGRAM_ADDRESS,
      accounts: [
         { address: input.accounts.owner, role: AccountRole.WRITABLE_SIGNER },
         { address: input.accounts.userVaultPda, role: AccountRole.WRITABLE },
         { address: input.accounts.userVaultAta, role: AccountRole.WRITABLE },
         { address: input.accounts.appAddress, role: AccountRole.READONLY },
         { address: input.accounts.sourceAta, role: AccountRole.WRITABLE },
         { address: input.accounts.mint, role: AccountRole.READONLY },
         { address: SYSTEM_PROGRAM_ADDRESS, role: AccountRole.READONLY },
         { address: input.accounts.tokenProgram, role: AccountRole.READONLY },
         { address: ASSOCIATED_TOKEN_PROGRAM_ADDRESS, role: AccountRole.READONLY },
      ] as const satisfies readonly AccountMeta[],
      data: encodeVaultIx({ kind: 'depositUserVault', amount: input.data.amount }),
   };
}

export function getUpdateUserVaultDelegateInstruction(input: {
   accounts: Readonly<{
      owner: Address;
      userVaultPda: Address;
      appAddress: Address;
      delegate: Address;
   }>;
   data: UpdateUserVaultDelegateInput;
}): Instruction {
   assertU32DelegateExpires(input.data.delegateExpires, 'updateUserVaultDelegate delegateExpires');
   return {
      programAddress: VAULT_PROGRAM_ADDRESS,
      accounts: [
         { address: input.accounts.owner, role: AccountRole.WRITABLE_SIGNER },
         { address: input.accounts.userVaultPda, role: AccountRole.WRITABLE },
         { address: input.accounts.appAddress, role: AccountRole.READONLY },
         { address: input.accounts.delegate, role: AccountRole.READONLY },
      ] as const satisfies readonly AccountMeta[],
      data: encodeVaultIx({
         kind: 'updateUserVaultDelegate',
         delegateExpires: input.data.delegateExpires,
      }),
   };
}

/** Matches `withdraw_user_vault.rs` account order (7 accounts). */
export function getWithdrawUserVaultInstruction(input: {
   accounts: Readonly<{
      owner: Address;
      userVaultPda: Address;
      userVaultAta: Address;
      appAddress: Address;
      destAta: Address;
      mint: Address;
      tokenProgram: Address;
   }>;
   data: WithdrawUserVaultInput;
}): Instruction {
   assertU64Amount(input.data.amount, 'withdraw amount');
   return {
      programAddress: VAULT_PROGRAM_ADDRESS,
      accounts: [
         { address: input.accounts.owner, role: AccountRole.WRITABLE_SIGNER },
         { address: input.accounts.userVaultPda, role: AccountRole.READONLY },
         { address: input.accounts.userVaultAta, role: AccountRole.WRITABLE },
         { address: input.accounts.appAddress, role: AccountRole.READONLY },
         { address: input.accounts.destAta, role: AccountRole.WRITABLE },
         { address: input.accounts.mint, role: AccountRole.READONLY },
         { address: input.accounts.tokenProgram, role: AccountRole.READONLY },
      ] as const satisfies readonly AccountMeta[],
      data: encodeVaultIx({ kind: 'withdrawUserVault', amount: input.data.amount }),
   };
}

/** Matches `withdraw_user_vault_native.rs`: owner receives lamports from the vault PDA (3 accounts). */
export function getWithdrawUserVaultNativeInstruction(input: {
   accounts: Readonly<{
      owner: Address;
      userVaultPda: Address;
      appAddress: Address;
   }>;
   data: WithdrawUserVaultNativeInput;
}): Instruction {
   assertU64Amount(input.data.amount, 'withdrawUserVaultNative amount');
   return {
      programAddress: VAULT_PROGRAM_ADDRESS,
      accounts: [
         { address: input.accounts.owner, role: AccountRole.WRITABLE_SIGNER },
         { address: input.accounts.userVaultPda, role: AccountRole.WRITABLE },
         { address: input.accounts.appAddress, role: AccountRole.READONLY },
      ] as const satisfies readonly AccountMeta[],
      data: encodeVaultIx({ kind: 'withdrawUserVaultNative', amount: input.data.amount }),
   };
}

export function getAppIxInstruction(input: {
   accounts: Readonly<{
      delegate: Address;
      owner: Address;
      userVaultPda: Address;
      appAddress: Address;
   }>;
   innerAccounts: readonly AccountMeta[];
   data: AppIxInput;
}): Instruction {
   if (!(input.data.innerInstructionData instanceof Uint8Array)) {
      throw new TypeError('getAppIxInstruction: innerInstructionData must be a Uint8Array');
   }
   const fixed: AccountMeta[] = [
      { address: input.accounts.delegate, role: AccountRole.WRITABLE_SIGNER },
      { address: input.accounts.owner, role: AccountRole.READONLY },
      { address: input.accounts.userVaultPda, role: AccountRole.WRITABLE },
      { address: input.accounts.appAddress, role: AccountRole.READONLY },
      { address: SYSVAR_CLOCK_ADDRESS, role: AccountRole.READONLY },
   ];
   return {
      programAddress: VAULT_PROGRAM_ADDRESS,
      accounts: [...fixed, ...input.innerAccounts],
      data: encodeVaultIx({
         kind: 'appIx',
         innerInstructionData: input.data.innerInstructionData,
      }),
   };
}

/**
 * Vault `CpiEntry` — 11 accounts (`program/src/instructions/cpi_entry.rs`).
 * Ends with instructions sysvar + clock sysvar. Writable flags follow `amountNative` / `amount`.
 */
export function getCpiEntryInstruction(input: {
   accounts: Readonly<{
      delegate: Address;
      owner: Address;
      userVaultPda: Address;
      userVaultAta: Address;
      appAddress: Address;
      lamportsDest: Address;
      destAta: Address;
      mint: Address;
      tokenProgram: Address;
   }>;
   data: CpiEntryInput;
}): Instruction {
   assertCpiEntryAmounts(input.data.amountNative, input.data.amount);
   const wNative = input.data.amountNative > 0n;
   const wSpl = input.data.amount > 0n;
   return {
      programAddress: VAULT_PROGRAM_ADDRESS,
      accounts: [
         { address: input.accounts.delegate, role: AccountRole.WRITABLE_SIGNER },
         { address: input.accounts.owner, role: AccountRole.READONLY },
         {
            address: input.accounts.userVaultPda,
            role: AccountRole.WRITABLE,
         },
         {
            address: input.accounts.userVaultAta,
            role: wSpl ? AccountRole.WRITABLE : AccountRole.READONLY,
         },
         { address: input.accounts.appAddress, role: AccountRole.READONLY },
         {
            address: input.accounts.lamportsDest,
            role: wNative ? AccountRole.WRITABLE : AccountRole.READONLY,
         },
         {
            address: input.accounts.destAta,
            role: wSpl ? AccountRole.WRITABLE : AccountRole.READONLY,
         },
         { address: input.accounts.mint, role: AccountRole.READONLY },
         { address: input.accounts.tokenProgram, role: AccountRole.READONLY },
         { address: SYSVAR_INSTRUCTIONS_ADDRESS, role: AccountRole.READONLY },
         { address: SYSVAR_CLOCK_ADDRESS, role: AccountRole.READONLY },
      ] as const satisfies readonly AccountMeta[],
      data: encodeVaultIx({
         kind: 'cpiEntry',
         amountNative: input.data.amountNative,
         amount: input.data.amount,
      }),
   };
}

/** Vault `CpiEntryNative` — 7 accounts (`program/src/instructions/cpi_entry_native.rs`; instructions + clock sysvars). */
export function getCpiEntryNativeInstruction(input: {
   accounts: Readonly<{
      delegate: Address;
      owner: Address;
      userVaultPda: Address;
      appAddress: Address;
      lamportsDest: Address;
   }>;
   data: CpiEntryNativeInput;
}): Instruction {
   assertU64Amount(input.data.amountNative, 'cpiEntryNative amountNative');
   return {
      programAddress: VAULT_PROGRAM_ADDRESS,
      accounts: [
         { address: input.accounts.delegate, role: AccountRole.WRITABLE_SIGNER },
         { address: input.accounts.owner, role: AccountRole.READONLY },
         { address: input.accounts.userVaultPda, role: AccountRole.WRITABLE },
         { address: input.accounts.appAddress, role: AccountRole.READONLY },
         { address: input.accounts.lamportsDest, role: AccountRole.WRITABLE },
         { address: SYSVAR_INSTRUCTIONS_ADDRESS, role: AccountRole.READONLY },
         { address: SYSVAR_CLOCK_ADDRESS, role: AccountRole.READONLY },
      ] as const satisfies readonly AccountMeta[],
      data: encodeVaultIx({
         kind: 'cpiEntryNative',
         amountNative: input.data.amountNative,
      }),
   };
}

export function getCloseVaultAtaInstruction(input: {
   accounts: Readonly<{
      owner: Address;
      userVaultPda: Address;
      appAddress: Address;
      userVaultAta: Address;
      destination: Address;
      mint: Address;
      tokenProgram: Address;
   }>;
}): Instruction {
   return {
      programAddress: VAULT_PROGRAM_ADDRESS,
      accounts: [
         { address: input.accounts.owner, role: AccountRole.WRITABLE_SIGNER },
         { address: input.accounts.userVaultPda, role: AccountRole.WRITABLE },
         { address: input.accounts.appAddress, role: AccountRole.READONLY },
         { address: input.accounts.userVaultAta, role: AccountRole.WRITABLE },
         { address: input.accounts.destination, role: AccountRole.WRITABLE },
         { address: input.accounts.mint, role: AccountRole.READONLY },
         { address: input.accounts.tokenProgram, role: AccountRole.READONLY },
      ] as const satisfies readonly AccountMeta[],
      data: encodeVaultIx({ kind: 'closeVaultAta' }),
   };
}

export function getCloseUserVaultInstruction(input: {
   accounts: Readonly<{
      owner: Address;
      userVaultPda: Address;
      appAddress: Address;
   }>;
}): Instruction {
   return {
      programAddress: VAULT_PROGRAM_ADDRESS,
      accounts: [
         { address: input.accounts.owner, role: AccountRole.WRITABLE_SIGNER },
         { address: input.accounts.userVaultPda, role: AccountRole.WRITABLE },
         { address: input.accounts.appAddress, role: AccountRole.READONLY },
      ] as const satisfies readonly AccountMeta[],
      data: encodeVaultIx({ kind: 'closeUserVault' }),
   };
}

export { SPL_TOKEN_PROGRAM_ADDRESS, SYSTEM_PROGRAM_ADDRESS, TOKEN_2022_PROGRAM_ADDRESS } from './constants';
